paparazzi/src/test/java/app/cash/paparazzi/PaparazziTest.kt

```diff
@@ -218,6 +218,27 @@ class PaparazziTest {
     assertThat(thrown).isTrue()
   }
 
+  @Test
+  fun preDrawOnEveryFrame() {
+    val order = mutableListOf<String>()
+    val view = object : View(paparazzi.context) {
+      override fun onAttachedToWindow() {
+        viewTreeObserver.addOnPreDrawListener {
+          order.add("predraw")
+        }
+      }
+
+      override fun onDraw(canvas: Canvas?) {
+        super.onDraw(canvas)
+        order.add("draw")
+      }
+    }
+
+    paparazzi.gif(view, fps = 4)
+
+    assertThat(order).isEqualTo(listOf("draw", "predraw", "predraw", "predraw"))
```

Comment:
> Noticed onPreDraw stopped getting callbacks for subsequent frames in an animation. 

I might be confused, so I'll ask a clarifying question.

This unit test isn't animating the view, so wouldn't pre-caching the bitmap and avoiding unnecessary predraws make sense?

Put another way, what if this view were animating?  Would it then show the predraws without the RenderParamsFlag change?

Also, why isn't there a "predraw" prior to the 1st "draw"?

Reply:
The preDrawListener is added to AttachInfo's ViewTreeObserver which is supposed to be invoked for every frame, it isn't a drawing call per-se rather a hey, we are about to render to the screen. 

Yes we are not animating the view which is why we don't expect subsequent draw's but the preDraw still should apply before each frame.

Because of the changes, [this](https://cs.android.com/android/platform/superproject/main/+/main:frameworks/base/core/java/android/view/ThreadedRenderer.java;l=700?q=ThreadedRenderer&ss=android%2Fplatform%2Fsuperproject%2Fmain) call causes draw to exectue, lower you can see [here](https://cs.android.com/android/platform/superproject/main/+/main:frameworks/base/core/java/android/view/ThreadedRenderer.java;l=742?q=ThreadedRenderer&ss=android%2Fplatform%2Fsuperproject%2Fmain) is where the preDraw is called before rendering the root's RenderingNode into the canvas. If preDraw causes an invalidation of the view, then [this](https://cs.android.com/android/platform/superproject/main/+/main:frameworks/base/core/java/android/view/ThreadedRenderer.java;l=745?q=ThreadedRenderer&ss=android%2Fplatform%2Fsuperproject%2Fmain) line will update the dirty list and render the final changes within the first frame.

If we updated the test to invalidate in preDraw we see the end results being

[draw, predraw, draw, predraw, draw, predraw, draw]

Notice 3 predraws but 4 draws. This is because invalidating the first frame in preDraw causes a 2nd draw phase before the first draw pass renders. Can update the test to invalidate to showcase that behavior. The pre iguana LayoutLib however it was as you'd be expecting in order `["predraw", "draw", "predraw", "predraw"]`

Reply:
> Yes we are not animating the view which is why we don't expect subsequent draw's but the preDraw still should apply before each frame.

Ah, got it.  TIL.

Btw, no need to add the `invalidate` to force the extra "draw"s, your explanation makes sense.